### PR TITLE
DBZ-2862 Documentation

### PR DESCRIPTION
**Pull request by vidourem**
Installation and configuration of OpenTelemetry to enable traceability

### CODE REVIEW

**Comment by jpechane** [comment]
@vjuranek @roldanbob Could you please review the docs? Thanks

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 1]
```diff
@@ -28,17 +18,89 @@ It is necessary to provide additional precautions when an application writes a r
...
+
+Then, in the directory containing this pom.xml, run the following command to download all dependencies (including transitives) to a specified directory (e.g. ./lib)
 
-[WARNING]
+[source,bash]
+----
+mvn dependency:copy-dependencies -DoutputDirectory=./lib
+----
+
+.Installing the SDK
```
Comment (vjuranek) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:79]:
```suggestion
==== Installing the SDK
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 2]
```diff
@@ -28,17 +18,89 @@ It is necessary to provide additional precautions when an application writes a r
...
-Debezium was tested with https://www.jaegertracing.io/[Jaeger] implementation.
+The tracing support is added to Debezium through the https://opentelemetry.io/docs/specs/otel/[OpenTelemetry specification].
+
+== Installation
+
+To enable traceability in Debezium, you need to install some OpenTelemetry API packages as well as the OpenTelemetry SDK on your Kafka Connect cluster.
+
+=== Manual installation
+
+.Installing the API
```
Comment (vjuranek) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:29]:
```suggestion
==== Installing the API
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 3]
```diff
@@ -28,17 +18,89 @@ It is necessary to provide additional precautions when an application writes a r
...
+mvn dependency:copy-dependencies -DoutputDirectory=./lib
+----
+
+.Installing the SDK
+
+Install the OpenTelemetry SDK on your Kafka connect cluster using the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent] for automatic instrumentation.
+
+=== Use the Debezium Kafka Connect image
+
+The OpenTelemetry API dependencies are already installed in https://quay.io/repository/debezium/connect[Debezium Kafka Connect] container image, you just need to install the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent]
```
Comment (vjuranek) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:85]:
```suggestion
The OpenTelemetry API dependencies are already installed in https://quay.io/repository/debezium/connect[Debezium Kafka Connect] container image.
The jar files are downloaded in separate directory and by default they are not included into the class path.
To add them into the class path, set environment variable `ENABLE_OTEL` to `true`.
You also need to install the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent]
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 4]
```diff
@@ -65,6 +127,13 @@ Optionally it is possible to enable tracing at the Kafka producer level.
 If enabled then when the message is being written to the Kafka broker the producer will extract Debezium's processing span context from the Kafka message headers, create a new child span and record information about the write to the broker.
 Then it injects the new span into the message headers so a consumer of the message can restore the trace and resume end-to-end tracing.
 
+The interceptor cannot propagate the traceability context if kafka instrumentation is enabled. The interceptor simply propagates the traceability context before delegating the instrumentation to the OpenTelemetry SDK.
```
Comment (vjuranek) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:130]:
```suggestion
The interceptor cannot propagate the traceability context if the Kafka instrumentation is enabled.
The interceptor simply propagates the traceability context before delegating the instrumentation to the OpenTelemetry SDK.
```

**Review by vjuranek — CHANGES_REQUESTED** [review]
Thanks for the PR @vidourem ! 
I think it's quite good, we can improve it later if user find it hard to follow.
I dropped a few suggestions (especially we should mention that the OTEL bits are added to the class path by setting `ENABLE_OTEL=true` in our container images).
Let's wait for @roldanbob if he has any further suggestions.

**Review by roldanbob — CHANGES_REQUESTED** [review]
I've completed an initial review of the content and suggested some changes. After you return and you have a chance to digest the review comments, let's do a second edit pass. Thanks!

**Comment by vidourem** [comment]
@vjuranek, @roldanbob,
I don't have my laptop with me, I tried a git client on smartphone but it doesn't do anything good. I will finish the documentation at the beginning of September, is it good for you?

**Comment by roldanbob** [comment]
> @vjuranek, @roldanbob, I don't have my laptop with me, I tried a git client on smartphone but it doesn't do anything good. I will finish the documentation at the beginning of September, is it good for you?

Not a problem for me. However, after 8 SEP, I will be on vacation until 19 SEP, so I'll be unable to provide any review/input during that time.

**Comment by vjuranek** [comment]
@vidourem sure, no problem. It would be good to have the documentation ready for 2.4 release which would contain switch to OTEL, but it's scheduled to end of Sept. so we have enough time (and in the worst case can do the proposed fixes ourselves). So please enjoy your holiday @vidourem (and @roldanbob as well).

**Comment by vidourem** [comment]
@vjuranek @roldanbob 
I just accepted all your suggestions, how could I not ;)
I'm not a native English speaker and I trust you completely with editing what I've written.
I'm going to take some time to check if everything is ok and I will try to find a tester on my side to check the procedure.

**Comment by roldanbob** [comment]
> @vjuranek @roldanbob I just accepted all your suggestions, how could I not ;) I'm not a native English speaker and I trust you completely with editing what I've written. I'm going to take some time to check if everything is ok and I will try to find a tester on my side to check the procedure.

Thanks for your efforts, @vjuranek. If you complete testing before the 7th, I can do a final check then.

**Review by vjuranek — APPROVED** [review]
Thanks a lot @vidourem , LGTM.
@roldanbob please do the final review.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 5]
```diff
@@ -9,36 +9,102 @@
...
-
 == Overview
 
 Observability is an important aspect of microservice-oriented applications.
 One of the key ingredients for observability is https://microservices.io/patterns/observability/distributed-tracing.html[distributed tracing].
 
 It is necessary to provide additional precautions when an application writes a record to a database that is later processed by Debezium.
 The active trace is effectively demarcated by the write to the database.
-If we want Debezium to join the larger scope application tracer we need to pass the trace metadata to Debezium.
+If we want {prodname} to join the larger scope application tracer, we need to pass the trace metadata to {prodname}.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:19]:
```suggestion
If you want to include {prodname} data in your application tracer, you must pass the trace metadata to {prodname}.
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 6]
```diff
@@ -9,36 +9,102 @@
...
-Debezium was tested with https://www.jaegertracing.io/[Jaeger] implementation.
+[source,bash]
+----
+mvn dependency:copy-dependencies -DoutputDirectory=./lib
+----
 
-[WARNING]
+==== Installing the SDK
+
+Install the OpenTelemetry SDK on your Kafka connect cluster using the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent] for automatic instrumentation.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:82]:
```suggestion
Install the OpenTelemetry SDK on your Kafka connect cluster by using the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent] for automatic instrumentation.
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 7]
```diff
@@ -9,36 +9,102 @@
...
+
+=== Use the Debezium Kafka Connect image
+
+The OpenTelemetry API dependencies are installed in the https://quay.io/repository/debezium/connect[Debezium Kafka Connect] container image.
+The JAR files are downloaded to a separate directory, and, by default, they are not included into the class path.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:87]:
```suggestion
The JAR files are downloaded to a separate directory.
By default, the JAR files are not added to the class path.
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 8]
```diff
@@ -9,36 +9,102 @@
...
+----
+mvn dependency:copy-dependencies -DoutputDirectory=./lib
+----
 
-[WARNING]
+==== Installing the SDK
+
+Install the OpenTelemetry SDK on your Kafka connect cluster using the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent] for automatic instrumentation.
+
+=== Use the Debezium Kafka Connect image
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:84]:
```suggestion
=== Use the {prodname} Kafka Connect image
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 9]
```diff
@@ -9,36 +9,102 @@
...
+
+The OpenTelemetry API dependencies are installed in the https://quay.io/repository/debezium/connect[Debezium Kafka Connect] container image.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:86]:
```suggestion
The OpenTelemetry API dependencies are installed in the https://quay.io/repository/debezium/connect[{prodname} Kafka Connect] container image.
```

**Review by roldanbob — CHANGES_REQUESTED** [review]
@vidourem A few more minor edits.

**Review by vjuranek — APPROVED** [review]

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 10]
```diff
@@ -9,36 +9,103 @@
...
+
+=== Use the {prodname} Kafka Connect image
+
+The OpenTelemetry API dependencies are installed in the https://quay.io/repository/debezium/connect[{prodname} Kafka Connect] container image.
+The JAR files are downloaded to a separate directory.
+By default, the JAR files are not added to the class path.
+To add the JAR files to the class path, set the environment variable `ENABLE_OTEL` to `true`.
+You must also install the https://opentelemetry.io/docs/instrumentation/java/automatic/[OpenTelemetry Java agent].
+
+=== Use Strimzi
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:92]: Similar question as in my preceding comment.  Let's be explicit in the heading about the purpose for using Strimzi in this context, i.e., Using Strimzi to [add the OpenTelemetry components?]

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 11]
```diff
@@ -9,36 +9,103 @@
...
+
+You can use the https://strimzi.io/[Strimzi] Kafka image, which provides all of the necessary OpenTelemetry components.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:94]: If this is an alternative method for providing the required components, let's use this opening sentence to explain that, for example, something like "Instead of using the {prodname} Kafka Connect image as described in the previous section, you can use obtain all of the OpenTelemetry components by deploying the https://strimzi.io/[Strimzi]) Kafka image."

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 12]
```diff
@@ -9,36 +9,103 @@
...
+
+If you want to review the various OpenTelemetry components, examine the compose file that is provided in the https://github.com/debezium/debezium-examples/tree/main/outbox[{prodname} Outbox example].
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:96]:
```suggestion
If you want to view the OpenTelemetry components that are available in the Strimzi image, examine the compose file that is provided in the https://github.com/debezium/debezium-examples/tree/main/outbox[{prodname} Outbox example].
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 13]
```diff
@@ -9,36 +9,103 @@
...
+
+=== Configure OpenTelemetry
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:98]:
```suggestion
=== Configuring OpenTelemetry
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 14]
```diff
@@ -62,9 +129,17 @@ When this SMT is invoked with a message then it will:
...
 
 Optionally it is possible to enable tracing at the Kafka producer level.
-If enabled then when the message is being written to the Kafka broker the producer will extract Debezium's processing span context from the Kafka message headers, create a new child span and record information about the write to the broker.
+If you enable tracing in the Kafka producer, when messages are written to the Kafka broker, the producer extracts the {prodname} processing span context from the Kafka message headers, creates a new child span, and then records information about the write operation to the broker.
 Then it injects the new span into the message headers so a consumer of the message can restore the trace and resume end-to-end tracing.
 
+The interceptor cannot propagate the traceability context if the Kafka instrumentation is enabled.
+The interceptor simply propagates the traceability context before delegating the instrumentation to the OpenTelemetry SDK.
+
+To enable end-to-end traceability:
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:138]:
```suggestion
==== Enabling end-to-end traceability
```

**Review by roldanbob — CHANGES_REQUESTED** [review]
@vidourem I reviewed this again after being away for over the past 10 days, and I realize that we could go further in clarifying a few things. See my comments. I also realize that I might be  wrong about how you can use the  Debezium KC image and Strimzi Kafka images to obtain the OpenTelemetry components. Let me know.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 15]
```diff
@@ -9,36 +9,103 @@
...
+mvn dependency:copy-dependencies -DoutputDirectory=./lib
+----
 
-[WARNING]
+==== Installing the SDK
+
+Install the OpenTelemetry SDK on your Kafka connect cluster by using the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent] for automatic instrumentation.
+
+=== Use the {prodname} Kafka Connect image
+
```
[the remark this answers is not in the document]
Comment (vidourem) [reply documentation/modules/ROOT/pages/integrations/tracing.adoc:97]: @roldanbob 

There are three different ways to install the Opentelemetry API and SDK. 

The first is completely free (manual installation) and allows the user to choose their own Kafka distribution, this is my case because I use the Confluent distribution. 

The second is a simplified way managed by Debezium with the Kafka connect image offered directly by Debezium. Personally I think this is useful for testing or prototyping only. 

The third highlights the Strimzi distribution which includes OpenTelemetry. It is intended for users who have chosen to use Strimzi on Kubernetes. personally I don't use it, but the Debezium team wanted plug and play integration with this distribution.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc (hunk 10 above)
Comment (vidourem) [reply documentation/modules/ROOT/pages/integrations/tracing.adoc]: see https://github.com/debezium/debezium/pull/4774#discussion_r1338854488

File: documentation/modules/ROOT/pages/integrations/tracing.adoc (hunk 11 above)
Comment (vidourem) [reply documentation/modules/ROOT/pages/integrations/tracing.adoc]: see https://github.com/debezium/debezium/pull/4774#discussion_r1338854488

**Comment by vidourem** [comment]
Hi @roldanbob , @vjuranek , @jpechane 

I have provided some clarification on the different installation methods. 

I also clarified the OpenTelemetry API Installation method for the case of manual installation, because after rereading I noticed an inaccuracy. 

Also see my comment https://github.com/debezium/debezium/pull/4774#discussion_r1338854488 

I hope it's clearer now. Do not hesitate to reformulate if necessary.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 16]
```diff
@@ -9,36 +9,116 @@
...
+Prefer this installation method if you want to use Docker for local development or testing.
+
+The OpenTelemetry API dependencies are installed in the https://quay.io/repository/debezium/connect[{prodname} Kafka Connect] container image.
+The JAR files are downloaded to a separate directory.
+By default, the JAR files are not added to the class path.
+To add the JAR files to the class path, set the environment variable `ENABLE_OTEL` to `true`.
+You must also install the https://opentelemetry.io/docs/instrumentation/java/automatic/[OpenTelemetry Java agent].
+
+=== Use Strimzi on Kubernetes [[use_trimzi]]
+https://strimzi.io/[Strimzi] provides a way to run an Apache Kafka cluster on Kubernetes.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:105]:
```suggestion
https://strimzi.io/[Strimzi] provides a way to deploy an Apache Kafka cluster on Kubernetes.
```

**Review by roldanbob — CHANGES_REQUESTED** [review]
Thanks, @vidourem. This is looking much better. I've proposed a few minor updates

**Comment by vidourem** [comment]
@roldanbob 

Accepted !

I also replaced OpenTracing with OpenTelemetry in the navigation menu and fixed an indentation issue

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 17]
```diff
@@ -9,36 +9,118 @@
...
+xref:open-telemetry-strimzi-installation[Strimzi installation]:: Presents a ready-to-use solution on Kubernetes that includes all of the necessary OpenTelemetry components.
+[id="open-telemetry-manual-installation"]
+=== Manual installation of OpenTelemetry
 
-[WARNING]
+A manual installation provides greater control and customization.
+
+==== Installing the OpenTelemetry API
+
+Install the following packages to in the Kafka Connect classpath:
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:39]:
```suggestion
Install the following packages to the Kafka Connect classpath:
```

Removes extra word

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 18]
```diff
@@ -9,36 +9,118 @@
...
+* https://mvnrepository.com/artifact/io.opentelemetry/opentelemetry-api/[opentelemetry-api]
+* https://mvnrepository.com/artifact/io.opentelemetry/opentelemetry-context/[opentelemetry-context]
+* https://mvnrepository.com/artifact/io.opentelemetry/opentelemetry-semconv/[opentelemetry-semconv]
+* https://mvnrepository.com/artifact/io.opentelemetry.instrumentation/opentelemetry-instrumentation-api/[opentelemetry-instrumentation-api]
+* https://mvnrepository.com/artifact/io.opentelemetry.instrumentation/opentelemetry-instrumentation-api-semconv/[opentelemetry-instrumentation-api-semconv]
+* https://mvnrepository.com/artifact/io.opentelemetry.instrumentation/opentelemetry-kafka-clients-2.6/[opentelemetry-kafka-clients-2.6]
+* https://mvnrepository.com/artifact/io.opentelemetry.instrumentation/opentelemetry-kafka-clients-common/[opentelemetry-kafka-clients-common]
+
+
+You can download them directly from the Maven repository or use the Maven dependency plugin to ensure version compatibility among the different OpenTelemetry packages.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:50]:
```suggestion
You can download the OpenTelemetry packages directly from the Maven repository, or retrieve the packages by using the Maven dependency plugin to ensure version compatibility.
```

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 19]
```diff
@@ -9,36 +9,118 @@
...
+</project>
+----
++
+2. In the directory that contains the `pom.xml` file, run the following command to download all dependencies (including transitives) to a specified directory (for example, `./lib`)
++
+[source,bash]
+----
+mvn dependency:copy-dependencies -DoutputDirectory=./lib
+----
++
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:90]:
```suggestion

```
Removes stray `+` sign.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 20]
```diff
@@ -9,36 +9,118 @@
...
+
+
+==== Installing the OpenTelemetry SDK
+After you install the API, install the OpenTelemetry SDK.
+Install the OpenTelemetry SDK on your Kafka connect cluster by using the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent] for automatic instrumentation.
+[id="open-telemetry-docker-installation"]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:96]:
```suggestion
Install the OpenTelemetry SDK on your Kafka connect cluster by using the https://opentelemetry.io/docs/instrumentation/java/automatic/[Java agent] for automatic instrumentation.

[id="open-telemetry-docker-installation"]
```

Please leave a blank line before the start of a new topic heading.

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 21]
```diff
@@ -9,36 +9,118 @@
...
+=== Using Docker to install OpenTelemetry 
+Prefer this installation method if you want to use Docker for local development or testing.
+
+The OpenTelemetry API dependencies are installed in the https://quay.io/repository/debezium/connect[{prodname} Kafka Connect] container image.
+The JAR files are downloaded to a separate directory.
+By default, the JAR files are not added to the class path.
+To add the JAR files to the class path, set the environment variable `ENABLE_OTEL` to `true`.
+You must also install the https://opentelemetry.io/docs/instrumentation/java/automatic/[OpenTelemetry Java agent].
+[id="open-telemetry-strimzi-installation"]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:105]:
```suggestion
You must also install the https://opentelemetry.io/docs/instrumentation/java/automatic/[OpenTelemetry Java agent].

[id="open-telemetry-strimzi-installation"]
```

Leave a blank line between topics.

**Review by roldanbob — CHANGES_REQUESTED** [review]
Corrects a few typos and adjusts some formatting for readability.
Thanks, @vidourem!

File: documentation/modules/ROOT/pages/integrations/tracing.adoc [hunk 22]
```diff
@@ -9,36 +9,119 @@
...
+== Installation
+
+To enable traceability in {prodname}, you must install the required OpenTelemetry API packages on your Kafka Connect cluster, together with the OpenTelemetry SDK.
+
+There are three main methods for installing OpenTelemetry on Kafka Connect, each having its advantages:
+
+xref:open-telemetry-manual-installation[Manual installation]:: Offers more control and customization.
+xref:open-telemetry-docker-installation[Docker installation]:: Provides easier integration.
+xref:open-telemetry-strimzi-installation[Strimzi installation]:: Presents a ready-to-use solution on Kubernetes that includes all of the necessary OpenTelemetry components.
+[id="open-telemetry-manual-installation"]
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/integrations/tracing.adoc:32]:
```suggestion

[id="open-telemetry-manual-installation"]
```

Missing new-line between topics

**Review by roldanbob — COMMENTED** [review]
LGTM. Remember to set off headers from surrounding text by placing newlines before the start of a topic and after the heading.

**Review by vjuranek — APPROVED** [review]
I'd consider squashing the commits, but the test LGTM. Thanks @vidourem !

**Review by roldanbob — APPROVED** [review]
Not sure if this was waiting on me. I see that the final tweak that I suggested was implemented, so from my perspective this is ready to go. 
For possible future productization I'll want to add downstream doc metatdata, but we can deal with that separately as necessary.

**Comment by jpechane** [comment]
@roldanbob @vidourem Squashed and applied, thanks!
